Assertion error during eta conversion / type unification (?)

```scala
trait Trait[X] {
    def f[C[_]](arg: C[X])
    val a = f[(Int, String)]
}
```

AE-e15bffbe1c7acc58f524073e849a753654ee312a

---

```scala
assertion failure for Tuple2[X] <:< Short, frozen = false
assertion failure for (Int, String)[X] <:< Short, frozen = false
exception occurred while typechecking test.scala
exception occurred while compiling test.scala
java.lang.AssertionError: assertion failed while compiling test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:15)
	at dotty.tools.dotc.core.Types$TypeAccumulator.foldArgs$1(Types.scala:4791)
	at dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:4802)
	at dotty.tools.dotc.core.Types$$anon$1.apply(Types.scala:128)
	at dotty.tools.dotc.core.Types$$anon$1.apply(Types.scala:110)
	at dotty.tools.dotc.core.Types$Type.testProvisional(Types.scala:133)
	at dotty.tools.dotc.core.Types$Type.isProvisional(Types.scala:106)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.record$1(SymDenotations.scala:1715)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.computeApplied$1(SymDenotations.scala:1791)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.recur$1(SymDenotations.scala:1794)
	at dotty.tools.dotc.core.SymDenotations$ClassDenotation.baseTypeOf(SymDenotations.scala:1850)
	at dotty.tools.dotc.core.Types$Type.baseType(Types.scala:925)
	at dotty.tools.dotc.core.TypeComparer.tryBaseType$1(TypeComparer.scala:656)
	at dotty.tools.dotc.core.TypeComparer.thirdTryNamed$1(TypeComparer.scala:454)
	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:471)
	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:428)
	at dotty.tools.dotc.core.TypeComparer.compareNamed$1(TypeComparer.scala:268)
	at dotty.tools.dotc.core.TypeComparer.firstTry$1(TypeComparer.scala:272)
	at dotty.tools.dotc.core.TypeComparer.recur(TypeComparer.scala:998)
	at dotty.tools.dotc.core.TypeComparer.compareAppliedType1$1(TypeComparer.scala:950)
	at dotty.tools.dotc.core.TypeComparer.fourthTry$1(TypeComparer.scala:692)
	at dotty.tools.dotc.core.TypeComparer.tryBaseType$1(TypeComparer.scala:663)
	at dotty.tools.dotc.core.TypeComparer.thirdTryNamed$1(TypeComparer.scala:454)
	at dotty.tools.dotc.core.TypeComparer.thirdTry$1(TypeComparer.scala:471)
	at dotty.tools.dotc.core.TypeComparer.secondTry$1(TypeComparer.scala:428)
```

--- a/compiler/src/dotty/tools/dotc/core/Substituters.scala
+++ b/compiler/src/dotty/tools/dotc/core/Substituters.scala
@@ -58,7 +58,7 @@ trait Substituters { this: Context =>
         val sym = tp.symbol
         var fs = from
         var ts = to
-        while (fs.nonEmpty) {
+        while (fs.nonEmpty && ts.nonEmpty) {
           if (fs.head eq sym) return ts.head
           fs = fs.tail
           ts = ts.tail
@@ -203,7 +203,7 @@ trait Substituters { this: Context =>
         val sym = tp.symbol
         var fs = from
         var ts = to
-        while (fs.nonEmpty) {
+        while (fs.nonEmpty && ts.nonEmpty) {
           if (fs.head eq sym)
             return ts.head match {
               case TypeBounds(lo, hi) => range(lo, hi)

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -4791,10 +4791,7 @@ object Types {
 
       case tp @ AppliedType(tycon, args) =>
         @tailrec def foldArgs(x: T, tparams: List[ParamInfo], args: List[Type]): T =
-          if (args.isEmpty) {
-            assert(tparams.isEmpty)
-            x
-          }
+          if (args.isEmpty || tparams.isEmpty) x
           else {
             val tparam = tparams.head
             val acc = args.head match {

--- a/tests/neg/i6063.scala
+++ b/tests/neg/i6063.scala
@@ -1,0 +1,12 @@
+object i0 {
+trait I1[i0, I1]
+def I1[I1[I1]](I1: I1[ ;  // error
+I1[I1]
+}
+
+object test {
+object I0 {
+trait i2[I1, i2]
+def i2[i2[i2]](i2: i2[ ]) = i2  // error
+i2[i2]}
+}

--- a/tests/neg/i6064.scala
+++ b/tests/neg/i6064.scala
@@ -1,0 +1,4 @@
+trait Trait[X] {
+    def f[C[_]](arg: C[X]): Int
+    val a = f[(Int, String)] // error
+}

